processor/src/main/java/com/linecorp/decaton/processor/SubscriptionStateListener.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor;
+
+import com.linecorp.decaton.processor.runtime.ProcessorSubscription;
+
+/**
+ * An interface to listen state changes of {@link ProcessorSubscription}.
+ * This interface is intended to be used for health check or integration testing purpose.
+ *
+ * For example, you can make sure that Decaton has started to process as follows:
+ * <pre>
+ * {@code
+ * CountDownLatch latch = new CountDownLatch(1);
+ * SubscriptionBuilder.newBuilder("my-subscription")
+ *                    // ...
+ *                    .stateListener(state -> {
+ *                        if (state == State.RUNNING) {
+ *                            latch.countDown();
+ *                        }
+ *                    }).buildAndStart();
+ * latch.await();
+ * // actual integration testing continues...
+ * }
+ * </pre>
+ */
+public interface SubscriptionStateListener {
+    /**
+     * Represents possible states that a {@link ProcessorSubscription} can be in.
+     * The expected state transition is:
+     *
+     * <pre>
+     *
+     *                 +-----+--------+
```

Comment:
I think this diagram is not exactly correct since the transition from `RUNNING` to `SHUTTING_DOWN` is possibly to happen?

Reply:
I omitted RUNNING to SHUTTING_DOWN because it only happens when something "unexpected" occurs e.g. OOOCC overflow.

....But, we know a case that ProcessorSubscription possibly die ungracefully when KafkaConsumer#committed throws an exception (which should be fixed later).
I'll add the transition.

Reply:
RUNNING to SHUTTING_DOWN exists in the diagram.

REBALANCING to SHUTTING_DOWN right?

Reply:
oops, no INITIALIZING to SHUTTING_DOWN.

Reply:
I imagined this diagram as usual expected transitions rather than all possible transitions, so I simplified the diagram.